
August 2015  
Modified September 2017, October 2018

Author: FBB

** reproduce the result in **
http://www.mdrc.org/sites/default/files/What%20Strategies%20Work%20for%20the%20Hard%20FR.pdf 

In [1]:
import os
import sys
import numpy as np
import pylab as pl

%pylab inline

Populating the interactive namespace from numpy and matplotlib


# NULL HYPOTHESIS: the % of former prisoners employed after release is the same or lower for candidates who participated in the program as for the control group, significance level p=0.05

** $H_0: P_0 - P_1 \geq$    0 **
    
** $H_a: P_0 - P_1 $< 0    **
    
    
** $\alpha$ = 0.05    **

** this is a TEST OF PROPORTIONS. we use the Binomial distribution since it is a yes/no (bernulli) test for each subject: the former inmate was or was not ever employed in a CEO transitional job (second row in the table above): **

** $P_0=0.035, P_1=0.701$**

In [2]:

# first thing: choose your significance threshold
alpha=0.05



# we like fractions better then percentages. as a rule of thumb, either use fractions or counts. 
# 0 for control group, 1 for program group

P_0 = 3.5 * 0.01 
P_1 = 70.1 * 0.01

if P_0 - P_1 >= 0:
    # we are done
    print ("the Null holds")
else:
    print ("we must assess the statistical significance")

#sample sizes 

n_0 = 409
n_1 = 564
    
#lets get the counts by multiplying by the sample size

Nt_0 = P_0 * n_0
Nt_1 = P_1 * n_1


we must assess the statistical significance


# WHICH  TEST?

samples, categorical data
 
 ** TWO OPTIONS z test, or chi-square test.  **

# START WITH Z TEST

** the z test compares the stanrard deviation of the expected distribution and the observed result. it tells you literally how many standard deviations from the tail an observation is, under the _assumption of normality_ **


must define the sample standard deviation (standard error)

In [3]:
#define the sample proportion first
sp = (P_0 * n_0 + P_1 * n_1) / (n_1 + n_0)
print (sp)

0.4210472764645426


** standard deviation of the sampling distribution the distribution is Binomial, the binomial stdev is **

(see a proof here!: http://stats.stackexchange.com/questions/29641/standard-error-for-the-mean-of-a-sample-of-binomial-random-variables!): 

$\sqrt{\frac{p(1 - p)}{n}}$

for 2 samples this becomes 

$\sqrt{ \frac{ \hat{p}(1 - \hat{p})} {n0} + \frac{ \hat{p}(1 - \hat{p})} {n1} }$

cfr: page 138 of Statistics in a Nutshell, eq. 5.12 and here http://stattrek.com/hypothesis-test/difference-in-proportions.aspx?Tutorial=AP

** $\hat{p} =\frac{p_0  n_0 + p_1  n_1}{n_0+n_1}$ **


Note that in the online version of Statistics in a Nutshell, $\bar{x}$  should be  $\hat{p}$!!

In [4]:
# I am goonna create a little one line function to calculate the standard error
# and to calculate p

p = lambda p0, p1, n0, n1: (p0 * n0 + p1 * n1) / (n0 + n1)
#standard error
se = lambda p, n0, n1: np.sqrt(p * (1 - p) * (1.0 / n0 + 1.0 / n1)) 

** z score: how many standard deviation away from the population parameter is my statistic? **

 $z=\frac{P_1-P_0}{\sigma}$

In [5]:
zscore = lambda p0, p1, s : (p0 - p1) / s
z_2y = zscore(P_1, P_0, se(p(P_0, P_1, n_0, n_1), n_0, n_1))
print (z_2y)


20.7697865408


note that using p0-p1 or p1-p0 at the numerator is equivalent because the standardizes normal value of z has mean 0 (see image below) so that we can use the absolute value of the z score, or equivalently look for $P[Z<z]$ if z is positive, and $P[Z>z]$ if z is negative.

* if $p<\alpha$ : reject H0 **

IMPORTANT!! note that this P in the bottom line of the table is not the p-value, but 

p-value = 1-P

In [6]:
## p-value for employment after 2 years: 
## since the largest number we read off the table for is (way) smaller 
## than the value for our statistic 
## our p-value will be smaller than it would be if calculated using 
## (e.g.) .9998 (and in fact using 1.0000 which is the largest number 
## in the table). Using 0.9998 is a **conservative** approach. 

p_2y = 1 - 0.9984


def report_result(p,a):
    print ('is the p value ' + 
           '{0:.2f} smaller than the critical value {1:.2f}?'.format(p,a))
    if p < a:
        print ("YES!")
    else: 
        print ("NO!")
    
    print ('the Null hypothesis is {}'.format(\
                            'rejected' if p < a  else 'not rejected') )

    
report_result(p_2y, alpha)

is the p value 0.00 smaller than the critical value 0.05?
YES!
the Null hypothesis is rejected


# Results of our test for the overall employment data:

Since we have rejected the null hypothesis, we are saying that former prisoners who participate in the CEO program have a higher rate of employment than those who do not participate in the program and that this difference is real, statistically significant, and likely not due to chance. 

# what if we used the values for where the former inmate was or was not "Convicted of a felony" (row 10) in the Recidivism (Years 1-3)?

## Null hypothesis:

For those individuals from the study who experienced recidivism within 3 years and were convicted of a crime, individuals who received the CEO employment program training (program/test grou) are just as or more likely to be convicted of a felony as those who did not receive the full CEO employment training (control group).

$H_0: P_0 - P_1 \leq$    0 

$H_a: P_0 - P_1 $ > 0    

 $P_0$ = The rate of felony convictions for the control group.
 
 $P_1$= The rate of felony convictions for the program/test group.

 look up data table and insert the appropriate values to get the appropriate result! you can use the functions I defined above, with different arguments. 

In [7]:
P_0=.117
P_1=.1

z_3y = zscore(P_1, P_0, se(p(P_0, P_1, n_0, n_1), n_0, n_1))
print(z_3y)

-0.846282982605


In [8]:
p_3y=1-.8023

report_result(p_3y, alpha)

is the p value 0.20 smaller than the critical value 0.05?
NO!
the Null hypothesis is not rejected


# Result of analysis:

Since we have not rejected the null, we must conclude that members of the test/program group were just as likely to have felony convictionions as members of the control group. In other words, the CEO program does not have a real effect on felony conviction rates for program participants.

# Now lets do it with the $\chi^2$ test

this analysis can also be done with the $\chi^2$ test, and the  $\chi^2$ distribution, 

see  flowcharts for how to choose a test and "Statistics In a Nutshell Chapter 4", or http://math.hws.edu/javamath/ryan/ChiSquare.html (if you are really just interested in the formula at face value)

** The chisq statistics tests the statistics calculated as : **

$\chi^2 = \sum_{i} \frac{(observation_i - expectation_i)^2}{expectation_i}$

 against a chi sq distribution.
 If we talk about sample fractions  that is 

$\chi^2 = \sum_i \frac{(f_{i,observed} - f_{i,expectated})^2}{f_{i,expected}}$

 Where _i_ indicates the sum over _each cell_.
 turns out this quantity is distributed according to a chi square distribution, so if i get the $\chi^2$ statistics i can compare it to the full chisq distribution and see how far in the tail it is

 The trickiest part (but not that tricky) is to figure out how to construct the table of values. please see Statistics In a Nutshell Chapter 4, for our data for example: Thisis called a CONTINGENCY TABLE

|                 |     success         | failure|    |               
|-----------------|:-------------------:|:-------------------:|---------------------------|
| test sample     | number of successes in test    | number of failures in test    | number members of test sample |
| control sample  | number of successes in control | number of failures in control | number members of control sample| 
|                 | total successes                |  total failures               | number of all members         |

|employed in subsadized job |     employed          | not employed     |                   
|---------------------------|:---------------------:|------------------|---------------------------|
| test sample               | $0.701*564$           | $0.299*564$      | 564                       |
| control sample            | $0.035*409$           | $0.965*4.09$     | 409                       |
|                           |                       |                  |                           |
| total                     | 409.679               |  562.912         | 973                       |

 for each cell the expected ratio is the product of the total of all rows and all columns, devided by the total

** FOR EACH CELL: **
**  expected_cell = $\frac{row~tot * col~tot}{total}$ **


This number must be compared to the chi sq distribution. 
You must calculate the number of degrees of freedom forthis experiment. 
Generally: DOF = Number of observations - number of Independent Variables

so here DOF = 1. Now you can look at the table below and draw conclusions about the rejection of the Null

432 is hella larger then 3.84

why am i mentioning 3.84?

how does the chi square statistic that you derived compare?

please  state what that means in terms of your Null hypothesis in a markdown cell below!

# !!!!!! TODO FOR YOU!
lets see what the chi sq statistics says about the conviction for fellonies (row 10)

# Response:

For the question above, regarding the note on 3.84, you have compared the Chi Square value of 432 from your sample data set to 3.84 because for one degree of freedom (which your sample data set has), a Chi Square value of 3.84 corresponds with an $\alpha$ level of .05. A Chi Square value greater than 3.84 (for one degree of freedom) would correspond with an $\alpha$ level less than .05. Therefore, we know that for the sample set the alpha level is less than .05 and we should reject the null hypothesis.

See below for my Chi Square statistic for convicted felons who received the complete employment program from CEO and the control group, which received minimal assistance:

In [9]:
# Calculations for table
ts = 568
cs = 409
t_p = .1
c_p = .117
t_no = 1 - t_p
c_no = 1 - c_p
t_tot = t_p*ts + t_no*ts
c_tot = c_p*cs + c_no*cs
y = t_p*ts + c_p*cs
n = t_no*ts + c_no*cs
tot = ts + cs
print(t_no)
print(c_no)
print(t_tot)
print(c_tot)
print(y)
print(n)
print(tot)

0.9
0.883
568.0
409.0
104.653
872.347
977


# Chi Square table for program and felony rate

|convicted of a felony      |     yes   | no        |Total                   
|---------------------------|-----------|-----------|----------------|
| test sample               |$.1*568   $|$.9*568   $|$568           $|
| control sample            |$.117*409 $|$.883*409 $|$409           $|
|                           |           |           |                |
| total                     |$104.653  $|$872.347  $|$977           $|

In [10]:
chi_pris = (tot*((t_p*ts*c_no*cs) - (t_no*ts*c_p*cs))**2)/(t_tot*c_tot*y*n)
print("Chi Square stat: " + str(chi_pris))

Chi Square stat: 0.7184939175052886


# Analysis of felony rate Chi Square statistic

Now that we have our Chi Square statistic of .718, we can go to our Chi Square distribution table and search for values in the row for one degree of freedom (since we have one degree of freedom in this case). Our Chi Square value of .718 does not match with an $\alpha$ level listed on the table, but we can deduce that the $\alpha$ is between .25 and .5 because the Chi Square statistics associated with those two $\alpha$ values are below and above .718.

Now that we know our Chi Square statistic corresponds with an $\alpha$ value somewhere between .25 and .5, we must conclude that we are not rejecting the null hypothesis since we set our $\alpha$ value at .05 for this study and the $\alpha$ value we derived from the study is greater than our threshold.

In conclusion, we cannot say with confidence that the CEO program has a real effect on the test/program group in terms of lowering their felony conviction rate. This conclusion matches our result from the z test.

In [11]:
# Calculations for table
ts = 564
cs = 409
t_p = .701
c_p = .035
t_no = 1 - t_p
c_no = 1 - c_p
t_tot = t_p*ts + t_no*ts
c_tot = c_p*cs + c_no*cs
y = t_p*ts + c_p*cs
n = t_no*ts + c_no*cs
tot = ts + cs
print(t_no)
print(c_no)
print(t_tot)
print(c_tot)
print(y)
print(n)
print(tot)

0.29900000000000004
0.965
564.0
409.0
409.679
563.321
973


In [12]:
chi_pris = (tot*((t_p*ts*c_no*cs) - (t_no*ts*c_p*cs))**2)/(t_tot*c_tot*y*n)
print("Chi Square stat: " + str(chi_pris))

Chi Square stat: 431.38403295126477
